### ISSUE TITLE
Runner unreliable with IPv4 only network

### ISSUE THREAD

**Issue by alsutton** [actions/runner#2739]
**Describe the bug**
TLDR; Actions are failing due to a lack of IPv4 fallback on connections to `api.github.com`

We have a LAN which has IPv6 and IPv4, but one of our ISPs only offers an IPv4 connection, so while our LAN is always IPv4 & IPv6 our upstream link is, at times, IPv4 only. If we run GitHub actions on Linux machines (Debian 11 or 12) without explicitly disabling IPv6 for the whole system our actions frequently fail with the following error when we're connected via the IPv4-only ISP;

```
[build](https://github.com/snappautomotive/monorepo/actions/runs/XXIDXX/job/XXIDXX#step:1:25)
Failed to download action 'https://api.github.com/repos/actions/checkout/tarball/XXIDXX'. Error: The request was canceled due to the configured HttpClient.Timeout of 100 seconds elapsing.
```

This will, on occasion, happen multiple times in a row causing a job to exhaust all retries and fail without running the job. This indicates, to us, that there is no fallback from IPv6 to IPv4 in place which would, ideally, happen after an IPv6 failure before a retry pause is initiated.

**To Reproduce**
* Install a Debian Linux 12 machine (Confirmed reproducable on D11, other Linux distributions may see the same issue).
* Configure an IPv6+IPv4 LAN with an IPv4 upstream link (router onwards)
* Install the GitHub action runner software
* Register the server as a self-hosted runner
* Leave running and watch for error reports.

**Expected behavior**
Actions would fail back to IPv4 when an IPv6 connection fails. 

## Runner Version and Platform
Confirmed on 2.306.1 and 2.307.1. We've not found a version which works reliably in this configuration

OS of the machine running the runner? Linux - Debian 12 and Debian 11

## What's not working?
Connectivity to `api.github.com` appears to be attempted using IPv6 in preference to IPv4, which, with an IPv4 only upstream link can cause repeated failures until a connection to the IPv4 address happens.

## Job Log Output
Error showing `HttpClient.Timeout` being triggered is above.

## Runner and Worker's Diagnostic Logs
Relevant information given above.

## Workaround
Disabling the IPv6 connectivity of the host by adding the following lines to  `/etc/sysctl.conf`;

```
# Disable IPv6
net.ipv6.conf.all.disable_ipv6 = 1
net.ipv6.conf.default.disable_ipv6 = 1
net.ipv6.conf.lo.disable_ipv6 = 1
```

This workaround is undesirable because it means the loss of connectivity to IPv6 resources on our LAN.

**Comment by alsutton** [comment]
We're seeing the same error even when IPv6 has been disabled using the workaround. This indicates, to us, that even when there is no IPv6 interface configured, the runner will still try to use an IPv6 address if it receives on from the DNS lookup.

**Comment by alsutton** [comment]
Thanks for the info, but we need to be able to show a well-supported supply chain for our software components, so switching to your action runner (or any action runner not "approved" by GitHub) isn't an option.

In terms of that bug, it's been open for over 5 years, so if it's not resolved by now, and there isn't an approved workaround, we'll probably switch to a different CI system.

**Comment by ovsds** [comment]
> Thanks for the info, but we need to be able to show a well-supported supply chain for our software components, so switching to your action runner (or any action runner not "approved" by GitHub) isn't an option.
> 
> In terms of that bug, it's been open for over 5 years, so if it's not resolved by now, and there isn't an approved workaround, we'll probably switch to a different CI system.

Having the same issue here, the only workaround that kinda worked was to prioratize ipv4 over ipv6 for resolutions except those that needed to be ipv6 adding next to `/etc/gai.conf`:
```
precedence {ipv6_priority_cidr} 110
precedence ::ffff:0:0/96  100
```
